Improve user experience with daemonized appoach

pkalever opened the pull request:
* cli: show better error msg when daemon is not running
* cli: provide a way to skip using daemon
* cli: show useful hint in header area of shell in daemonized mode

pkalever commented:
@maurizio-lombardi @gonzoleeman @lxbsz please review, thanks!

lxbsz commented:
@pkalever 

The --skip-daemon seems not working well.

```
[targetcli-fb]# targetcli --skip-deamon
[Errno 111] Connection refused
Currently auto_use_daemon is true, hence please make sure targetclid daemon is running ...
(or)
Incase if you wish to turn auto_use_daemon to false then run '#targetcli --skip-daemon'
[targetcli-fb]# targetcli --skip-deamon
[Errno 111] Connection refused
Currently auto_use_daemon is true, hence please make sure targetclid daemon is running ...
(or)
Incase if you wish to turn auto_use_daemon to false then run '#targetcli --skip-daemon'

```

I just stopped the targetclid service just before the above. Did I miss something here ?

pkalever commented:
> I just stopped the targetclid service just before the above. Did I miss something here ?

Yes you did, you had a typo

> [targetcli-fb]# targetcli --skip-deamon

s/deamon/daemon/

lxbsz commented:
> > I just stopped the targetclid service just before the above. Did I miss something here ?
> 
> Yes you did, you had a typo
> 
> > [targetcli-fb]# targetcli --skip-deamon
> 
> s/deamon/daemon/

Ah, such a stupid typo :-)

Test it again and it works for me:

```
[targetcli-fb]# targetcli --skip-daemon
Parameter auto_use_daemon is now 'false'.
[targetcli-fb]# targetcli ls
o- / ......................................................................................................................... [...]
  o- backstores .............................................................................................................. [...]
  | o- block .................................................................................................. [Storage Objects: 0]
  | o- fileio ................................................................................................. [Storage Objects: 0]
  | o- pscsi .................................................................................................. [Storage Objects: 0]
  | o- ramdisk ................................................................................................ [Storage Objects: 0]
  | o- user:fbo ............................................................................................... [Storage Objects: 0]
  | o- user:glfs .............................................................................................. [Storage Objects: 2]
  |   o- block0 ........................... [repvol@localhost/block-store/ab08964d-ac58-4d5a-8f16-4ed4db5b73be (100.0MiB) activated]
  |   | o- alua ................................................................................................... [ALUA Groups: 3]
  |   |   o- default_tg_pt_gp ....................................................................... [ALUA state: Active/optimized]
  |   |   o- glfs_tg_pt_gp_ano .................................................................. [ALUA state: Active/non-optimized]
  |   |   o- glfs_tg_pt_gp_ao ....................................................................... [ALUA state: Active/optimized]
  |   o- block1 ........................... [repvol@localhost/block-store/c42d7d43-72d7-4bc5-a233-094359db933c (100.0MiB) activated]
  |     o- alua ................................................................................................... [ALUA Groups: 3]
  |       o- default_tg_pt_gp ....................................................................... [ALUA state: Active/optimized]
  |       o- glfs_tg_pt_gp_ano .................................................................. [ALUA state: Active/non-optimized]
  |       o- glfs_tg_pt_gp_ao ....................................................................... [ALUA state: Active/optimized]
  o- iscsi ............................................................................................................ [Targets: 2]
  | o- iqn.2016-12.org.gluster-block:ab08964d-ac58-4d5a-8f16-4ed4db5b73be ................................................ [TPGs: 1]
  | | o- tpg1 .................................................................................................. [gen-acls, no-auth]
  | |   o- acls .......................................................................................................... [ACLs: 0]
  | |   o- luns .......................................................................................................... [LUNs: 1]
  | |   | o- lun0 ................................................................................. [user/block0 (glfs_tg_pt_gp_ao)]
  | |   o- portals .................................................................................................... [Portals: 1]
  | |     o- 192.168.195.164:3260 ............................................................................................. [OK]
  | o- iqn.2016-12.org.gluster-block:c42d7d43-72d7-4bc5-a233-094359db933c ................................................ [TPGs: 1]
  |   o- tpg1 .................................................................................................. [gen-acls, no-auth]
  |     o- acls .......................................................................................................... [ACLs: 0]
  |     o- luns .......................................................................................................... [LUNs: 1]
  |     | o- lun0 ................................................................................. [user/block1 (glfs_tg_pt_gp_ao)]
  |     o- portals .................................................................................................... [Portals: 1]
  |       o- 192.168.195.164:3260 ............................................................................................. [OK]
  o- loopback ......................................................................................................... [Targets: 0]
  o- vhost ............................................................................................................ [Targets: 0]
  o- xen-pvscsi ....................................................................................................... [Targets: 0]
[targetcli-fb]# 

```

With this patch set it much better for the use experience.

Thanks.

lxbsz commented:
After the targetclid service is restarted, the following command must be executed at least twice:

`# targetcli set global auto_use_daemon=true`

```
[root@rhel3 gluster-block]# targetcli get global auto_use_daemon
------- [['get', 'global', 'auto_use_daemon']]

[Errno 111] Connection refused
Currently auto_use_daemon is true, hence please make sure targetclid daemon is running ...
(or)
Incase if you wish to turn auto_use_daemon to false then run '#targetcli --skip-daemon'
[root@rhel3 gluster-block]# targetcli --skip-daemon
------- [['--skip-daemon']]

Parameter auto_use_daemon is now 'false'.
[root@rhel3 gluster-block]# targetcli get global auto_use_daemon
------- [['get', 'global', 'auto_use_daemon']]

auto_use_daemon=false 
[root@rhel3 gluster-block]# systemctl start targetclid
[root@rhel3 gluster-block]# targetcli get global auto_use_daemon
------- [['get', 'global', 'auto_use_daemon']]

auto_use_daemon=false 
[root@rhel3 gluster-block]# targetcli set global auto_use_daemon=true
------- [['set', 'global', 'auto_use_daemon=true']]

Parameter auto_use_daemon is now 'true'.
[root@rhel3 gluster-block]# targetcli get global auto_use_daemon
------- [['get', 'global', 'auto_use_daemon']]

auto_use_daemon=false 
[root@rhel3 gluster-block]# targetcli set global auto_use_daemon=true
------- [['set', 'global', 'auto_use_daemon=true']]

Parameter auto_use_daemon is now 'true'.
[root@rhel3 gluster-block]# targetcli get global auto_use_daemon
------- [['get', 'global', 'auto_use_daemon']]

auto_use_daemon=true 
[root@rhel3 gluster-block]# 

```

lxbsz commented:
@pkalever 

This is because that after we just `systemctl start targetclid` service and at the same time the 'auto_use_daemon" is still false. And the `targetcli set global auto_use_daemon=true` will run in the interactive mode, so the copy of the "shell.prefs['auto_use_daemon']" in the daemon is skipped.

We need to check the argv and if it is "set global auto_use_daemon=true", we should run into the daemon too.

```
227     skip_daemon=False
228     if len(sys.argv) > 1:
229         usage_version(sys.argv[1])
230         if sys.argv[1] in ("tcp", "--tcp", "-t"):
231             use_daemon = True
232         elif sys.argv[1] in ("skip-daemon", "--skip-daemon"):
233             skip_daemon=True
234 
235     if `auto_use_daemon_is_true` or use_daemon and not skip_daemon:                                                                                                        
236         call_daemon(shell, get_arguments(shell).encode())
237         # does not return
238 

```

lxbsz commented:
> @pkalever
> 
> This is because that after we just `systemctl start targetclid` service and at the same time the 'auto_use_daemon" is still false. And the `targetcli set global auto_use_daemon=true` will run in the interactive mode, so the copy of the "shell.prefs['auto_use_daemon']" in the daemon is skipped.
> 
> We need to check the argv and if it is "set global auto_use_daemon=true", we should run into the daemon too.
> 
> Something like by adding `auto_use_daemon_is_true `:
> 
> ```
> 227     skip_daemon=False
> 228     if len(sys.argv) > 1:
> 229         usage_version(sys.argv[1])
> 230         if sys.argv[1] in ("tcp", "--tcp", "-t"):
> 231             use_daemon = True
> 232         elif sys.argv[1] in ("skip-daemon", "--skip-daemon"):
> 233             skip_daemon=True
> 234 
> 235     if auto_use_daemon_is_true or use_daemon and not skip_daemon:                                                                                                        
> 236         call_daemon(shell, get_arguments(shell).encode())
> 237         # does not return
> 238 
> ```

The above is not correct fixing, the root cause is that in the targetcli none daemon mode the shell.prefs from the ~/.targetcli/  are not consistent with which from targetclid daemon. The targetclid daemon is reading it from the cache not from the file every time, because after the targetclid daemon restart the data will be as expected.

lxbsz commented:
Sorry for confusing from the above comments.

1, The targetcli cmd in interactive mode will update the new options to self.shell.prefs and then save them into ~/.targetcli/prefs.bin every time before exiting. So this will be okay without any problem.

2, But when the targetcli cmd goes into the targetclid mode, every time we are trying to 'get auto_use_daemon', it will read it from the targetclid's self.shell.prefs cache instead, which won't load prefs from ~/.targetcli/prefs.bin every time. So any update in the Step#1 won't be seen here. With the following simple fixing it works for me:

```
diff --git a/configshell/shell.py b/configshell/shell.py
index bb4dd32..9135afe 100644
--- a/configshell/shell.py
+++ b/configshell/shell.py
@@ -808,6 +809,11 @@ class ConfigShell(object):
                 pparams = ['.']
 
         try:
+            self.prefs.load()
+        except IOError:
+            self.log.warning("Could not load preferences file %s."
+                             % self._prefs_file)
+        try:
             target = self._current_node.get_node(path)
         except ValueError as msg:
             raise ExecutionError(str(msg))
```

Thanks.

pkalever commented:
@lxbsz updated the PR with the fix, please help review. Thanks!

lxbsz commented:
> @lxbsz updated the PR with the fix for issue you have pointed above, it will be clean to load/read the prefs before running any new command within the daemon, and hence changes shouldn't go in configshell.
> 
> This fix in targetclid [88e5b95](https://github.com/open-iscsi/targetcli-fb/commit/88e5b954e72827eaa7157fc792440719fb688a99) should nail it down, please help review. 

@pkalever 
Test again and this works and is perfect for me.
Thanks

gonzoleeman commented:
I will review this evening ...

gonzoleeman commented:
So what is the context of "--skip_daemon"? Since there is no man page, and "targetcli --help" is of little value, I am left to guess that it means "don't use the daemon for this single invocation of targetcli". But evidently, based on the message, it is disabling the global variable that says I want to use the daemon?

```
root@linux-tw7(1):/home/lduncan# targetcli --skip-daemon     
Parameter auto_use_daemon is now 'false'.
root@linux-tw7(1):/home/lduncan# targetcli --skip-daemon abc def
Parameter auto_use_daemon is now 'false'.
```

And it seems to give this message every time, i.e. even when I call it two times in a row? That seems like overkill. Plus, it ignores other command line arguments?

So what exactly is "--skip-daemon" supposed to do? If it is supposed to handle the remaining command-line arguments without the daemon, it is failing, and the help info is incorrect. If it supposed to clear the global flag and exit, the help info needs to be updated.

pkalever commented:
@gonzoleeman 
'  --skip-daemon         skip using daemon'
Yes, I agree with you, the help needs some more clarification.

How about:
'--skip-daemon         turnoff the global auto use daemon flag' ?


@maurizio-lombardi @gonzoleeman @lxbsz I'm really bad at naming the options, please help choosing the names for --tcp option and --skip-daemon option, because once targetclid is release in production ready its very hard to change them.

Also, I thought of writing man page for targetclid daemon, but the daemon doesn't take any options now other than --help and --version, will it be good any easy to club the details with exiting targetcli man page ?


Thanks

pkalever commented:
V3:

> Also, I thought of writing man page for targetclid daemon, but the daemon doesn't take any options now other than --help and --version, will it be good any easy to club the details with exiting targetcli man page ?

* Clubbing targetclid details in the targetcli man page looked clumsy for me, hence added a new targetclid.8 man entry.

* fixed the help description for '--skip-daemon'

* Fixed doc in the systemd units


Please help review. Thanks!

gonzoleeman commented:
You could easily add targetclid to the targetcli man page IMHO. Also, addressing this could and probably should be a new separate issue.

lxbsz commented on targetclid.8:
```diff
@@ -0,0 +1,87 @@
+.BR targetcli (8),
+.BR targetctl (8),
+.BR tcmu-runner (8),
+.BR gluster-block (8)
```
We may need to update the "SEE ALSO" in the targetcli.8 too.

SEE ALSO
       targetclid(8), targetctl(8), tcmu-runner(8)

pkalever replied:
Yep, will add it.

lxbsz commented on targetclid.8:
```diff
@@ -0,0 +1,87 @@
+.TP
+If you would like to use the daemonized approach as default method then,
+.br
+$ targetcli set global auto_use_daemon=true
```
Not sure should we also mention that there has one daemon mode stuff in the targetcli.8 ?

maurizio-lombardi commented:
> How about:
> '--skip-daemon turnoff the global auto use daemon flag' ?
> 
> @maurizio-lombardi @gonzoleeman @lxbsz I'm really bad at naming the options, please help
> choosing the names for --tcp option and --skip-daemon option,

Considering that --skip-daemon is used to turn off the global auto_use_daemon setting,
I would rename it to --disable-daemon

pkalever commented:
@maurizio-lombardi '--disable-daemon' looks cool for disabling daemon mode. Just opted the patches to the same. Thanks!
